presto-main/src/main/java/com/facebook/presto/server/TaskInfoResource.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server;
+
+import com.facebook.presto.execution.QueryManager;
+import com.facebook.presto.execution.StageInfo;
+import com.facebook.presto.execution.TaskId;
+import com.facebook.presto.execution.TaskInfo;
+import com.facebook.presto.spi.QueryId;
+import com.google.inject.Inject;
+
+import javax.annotation.security.RolesAllowed;
+import javax.ws.rs.GET;
+import javax.ws.rs.NotFoundException;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+@Path("v1/taskInfo")
+@RolesAllowed("ADMIN")
+public class TaskInfoResource
+{
+    private final QueryManager queryManager;
+
+    @Inject
+    public TaskInfoResource(QueryManager queryManager)
+    {
+        this.queryManager = requireNonNull(queryManager, "queryManager is null");
+    }
+
+    @GET
+    @Path("{taskId}")
+    public Optional<TaskInfo> getTaskInfo(@PathParam("taskId") TaskId taskId)
+            throws NotFoundException
+    {
+        QueryId queryId = taskId.getQueryId();
+        try {
+            Optional<StageInfo> stageInfo = queryManager.getFullQueryInfo(queryId).getOutputStage();
+
+            if (stageInfo.isPresent()) {
+                Optional<TaskInfo> taskInfo = stageInfo.get().getAllStages().stream()
```

Comment:
getAllStages does a preOrder traversal on a constructed tree
```
    public List<StageInfo> getAllStages()
    {
        return ImmutableList.copyOf(forTree(StageInfo::getSubStages).depthFirstPreOrder(this));
    }
```
Can we directly call into `getSubStages` for this processing

Reply:
I looked into this. It seems like calling getSubStages will only return the first layer, so we won't be able to get all the stages we need to find the stage associated with the TaskId. Let me know if you think there's a way around it!

Reply:
Ah. So here is what we can do to optimize this part.
Instead of traversing and collecting the entire structure and then filtering, do a preorder on the tree and filter as you go along. 
Also it might be a good idea to get a sense of the tree we are traversing to check if bfs is a better fit.
If this method is going to be rarely used don't bother with this optimization. If this will be on the hot path then might be a good optimization to add in.

cc @tdcmeehan

Reply:
Because this will be used to service ad-hoc UI requests for individual tasks, my estimation is that this falls in the rarely used (but important when it is used) category of code.

Reply:
@chi-tsai if you think this is a simple enough change go ahead and do it else we can revisit it later . Just add a TODO comment.

Reply:
Updated with the change you mentioned